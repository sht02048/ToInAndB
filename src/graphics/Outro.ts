import Sound from "../utils/Sound";
import Renderer from "./Renderer";

import FPS from "../constants/fps";
import { IMAGE, AUDIO } from "../constants/path";

class Outro {
  #fps = FPS;
  #outSpeed = 1;
  #titleWidth = 795;
  #messageWidth = 772;
  #haltFrame = 16 * this.#fps;
  #hasMusicStarted = false;

  private title: Renderer;
  private message: Renderer;
  private outroMusic: Sound;
  private canvasWidth: number;
  private x: number;
  private y: number;

  public isOut: boolean;
  public shouldBeDisplayed: boolean;

  constructor() {
    this.title = new Renderer(IMAGE.OUTRO_TITLE);
    this.message = new Renderer(IMAGE.OUTRO_MESSAGE);
    this.outroMusic = new Sound(AUDIO.OUTRO);

    this.canvasWidth = this.title.canvasWidth;
    this.x = this.canvasWidth / 2;
    this.y = 0;

    this.isOut = false;
    this.shouldBeDisplayed = false;
  }

  update() {
    if (this.y + this.title.height + this.message.height + 600 < 0) {
      this.isOut = true;
    }

    if (!this.#hasMusicStarted) {
      this.playOutroMusic();
    }

    if (this.#haltFrame > 0) {
      this.#haltFrame -= 1;
      return;
    }

    this.y -= this.#outSpeed;
  }

  render() {
    this.title.render(this.x - this.#titleWidth / 2, this.y + 100);
    this.message.render(this.x - this.#messageWidth / 2, this.y + 400);
  }

  playOutroMusic() {
    this.outroMusic.playAudio();
    this.#hasMusicStarted = true;
  }

  pauseOutroMusic() {
    this.outroMusic.pauseAudio();
  }
}

export default Outro;
