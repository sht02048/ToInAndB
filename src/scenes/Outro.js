import Sound from "../utils/Sound";
import { IMAGE, AUDIO } from "../constants/path";
import Renderer from "../graphics/Renderer";
import MODIFIER from "../constants/modifier";

class Outro {
  #fps = MODIFIER.SPEED === 1 ? 120 : 60;
  #outSpeed = 1 * MODIFIER.SPEED;
  #titleWidth = 795;
  #messageWidth = 772;
  #haltFrame = 15.5 * this.#fps;
  #hasMusicStarted = false;

  constructor() {
    this.title = new Renderer(IMAGE.OUTRO_TITLE);
    this.message = new Renderer(IMAGE.OUTRO_MESSAGE);

    this.outroMusic = new Sound(AUDIO.OUTRO);

    this.canvasWidth = this.title.canvasWidth;

    this.x = this.canvasWidth / 2;
    this.y = 0;

    this.isOut = false;
    this.shouldBeDisplayed = false;
  }

  update() {
    if (this.y + this.title.height + this.message.height + 600 < 0) {
      this.isOut = true;
    }

    if (!this.#hasMusicStarted) {
      this.playOutroMusic();
    }

    if (this.#haltFrame > 0) {
      this.#haltFrame -= 1;
      return;
    }

    this.y -= this.#outSpeed;
  }

  render() {
    this.title.render(this.x - this.#titleWidth / 2, this.y + 100);
    this.message.render(this.x - this.#messageWidth / 2, this.y + 400);
  }

  playOutroMusic() {
    this.outroMusic.playAudio();
    this.#hasMusicStarted = true;
  }

  pauseOutroMusic() {
    this.outroMusic.pauseAudio();
  }
}

export default Outro;
