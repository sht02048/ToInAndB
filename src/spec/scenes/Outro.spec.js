import Outro from "../../scenes/Outro";

describe("Outro", () => {
  let haltFrame = 1920;
  const mainCanvas = document.createElement("canvas");
  const introCanvas = document.createElement("canvas");
  mainCanvas.id = "main-canvas";
  introCanvas.id = "intro-canvas";

  document.body.appendChild(mainCanvas);
  document.body.appendChild(introCanvas);

  let outro;

  beforeEach(() => {
    outro = new Outro();
  });

  it("should have proper functions", () => {
    expect(outro.update).to.be.a("function");
    expect(outro.render).to.be.a("function");
    expect(outro.playOutroMusic).to.be.a("function");
    expect(outro.pauseOutroMusic).to.be.a("function");
  });

  it("should render", () => {
    outro.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should update", () => {
    const originalY = outro.y;

    while (haltFrame >= 0) {
      outro.update();
      haltFrame -= 1;
    }

    expect(originalY).to.not.equal(outro.y);
  });
});
