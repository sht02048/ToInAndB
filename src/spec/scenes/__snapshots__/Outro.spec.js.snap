// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Outro > should render 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -247.5,
      "dy": 100,
      "img": <img
        src="/images/outro_title.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -236,
      "dy": 400,
      "img": <img
        src="/images/outro_message.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
